edb/edgeql-parser/edgeql-parser-python/src/parser.rs

```diff
@@ -0,0 +1,164 @@
+    let (spec, productions) = load_spec(py, parser_name.to_string(py)?.as_ref())?;
+
+    let tokens = downcast_tokens(py, tokens)?;
+
+    let context = parser::Context::new(spec);
+    let (cst, errors) = parser::parse(&tokens, &context);
+
+    let cst = cst.map(|c| to_py_cst(c, py)).transpose()?;
+
+    let errors = errors
+        .into_iter()
+        .map(|e| parser_error_into_tuple(py, e))
+        .collect::<Vec<_>>();
+    let errors = PyList::new(py, &errors);
+
+    let res = ParserResult::create_instance(py, cst.into_py_object(py), errors)?;
+
+    Ok((res, productions).into_py_object(py))
+}
+
+py_class!(pub class CSTNode |py| {
+    data _production: PyObject;
+    data _terminal: PyObject;
+
+    def production(&self) -> PyResult<PyObject> {
+        Ok(self._production(py).clone_ref(py))
+    }
+    def terminal(&self) -> PyResult<PyObject> {
+        Ok(self._terminal(py).clone_ref(py))
+    }
+});
+
+py_class!(pub class Production |py| {
+    data _id: PyInt;
+    data _args: PyList;
+
+    def id(&self) -> PyResult<PyInt> {
+        Ok(self._id(py).clone_ref(py))
+    }
+    def args(&self) -> PyResult<PyList> {
+        Ok(self._args(py).clone_ref(py))
+    }
+});
+
+py_class!(pub class Terminal |py| {
+    data _text: PyString;
+    data _value: PyObject;
+    data _start: u64;
+    data _end: u64;
+
+    def text(&self) -> PyResult<PyString> {
+        Ok(self._text(py).clone_ref(py))
+    }
+    def value(&self) -> PyResult<PyObject> {
+        Ok(self._value(py).clone_ref(py))
+    }
+    def start(&self) -> PyResult<u64> {
+        Ok(*self._start(py))
+    }
+    def end(&self) -> PyResult<u64> {
+        Ok(*self._end(py))
+    }
+});
+
+static mut PARSER_SPECS: Option<HashMap<String, (parser::Spec, PyObject)>> = None;
+
+pub fn init_module() {
+    unsafe {
+        PARSER_SPECS = Some(HashMap::new());
+    }
+}
+
+fn downcast_tokens<'a>(py: Python, token_list: PyObject) -> PyResult<Vec<parser::Terminal>> {
+    let tokens = PyList::downcast_from(py, token_list)?;
+
+    let mut buf = Vec::with_capacity(tokens.len(py));
+    for token in tokens.iter(py) {
+        let token = OpaqueToken::downcast_from(py, token)?;
+        let token = token.inner(py);
+
+        buf.push(parser::Terminal::from_token(token));
+    }
+    Ok(buf)
+}
+
+fn load_spec(py: Python, parser_name: &str) -> PyResult<&'static (parser::Spec, PyObject)> {
+    let parser_specs = unsafe { PARSER_SPECS.as_mut().unwrap() };
+    if !parser_specs.contains_key(parser_name) {
+        let parser_mod = py.import("edb.edgeql.parser.parser")?;
+
+        let process_spec = py.import("edb.edgeql.parser")?.get(py, "process_spec")?;
+
+        let parser_cls = parser_mod.get(py, parser_name)?;
+        let parser = parser_cls.call(py, PyTuple::new(py, &[]), None)?;
+
+        let res = process_spec.call(py, (parser,), None)?;
+        let res = PyTuple::downcast_from(py, res).expect("process_spec to return a tuple");
```

Comment:
Yes it will. As a general rule, I resort to panic when something is obviously a bug and was not handled properly, which is the case here. Just to make sure that the bug is found during testing and is not handled by some exception handler somewhere.

But since we are using python it would be nicer if it threw an exception, yes